pkg/apis/sessions/session_state_test.go

```diff
@@ -186,6 +186,17 @@ func TestEncodeAndDecodeSessionState(t *testing.T) {
 			IDToken:     "IDToken.12349871293847fdsaihf9238h4f91h8fr.1349f831y98fd7",
 			ExpiresOn:   &expires,
 		},
+		"With groups": {
```

Comment:
An important test scenario I think we are missing with how these are constructed (struct -> encoded -> decode -> struct):

Does this handle an encoded 6.0 session state (which isn't aware of groups). Coming into to this branch which has groups? I'm assuming the MessagePack unmarshall will just return empty groups and this is fine? (similar to the LegacyV5DecodeSessionState for JSON encoded sessions that are passing just fine).

If unit testing this scenario is arduous, a quick live check of getting a session cookie via a v6.0.0 and then updating your server (with no option changes) and confirming you aren't logged out will suffice.

Related: I think we should document in important notes a rollout path for users who want this without logging out users. This will require a cookie refresh or cookie expiration & the user will need to have gone through an auth or refresh to get the groups into their session. We should make a note for sysadmins to give a rollout window so the group data can propogate onto sessions before setting `allowed-group` otherwise it will logout existing users if done too soon.

Reply:
I am going to opt for a manual test for this and then look at what I need to document for the "rollout path".

Reply:
@NickMeves did the test to go from v6.0.0 with a cookie to my version and it works without logging me out. Final task is the work for the rollout plan but from my end that appeared seamless, I did the following in my test:

1. Deploy v6.0.0 get a cookie
2. Upgrade to my version
3. Was not logged out
4. Deployed config change specifying groups
5. Was logged out and sent back to the login page where I logged in again and everything worked as expected

Curious on the steps that the plan you detail above would entail?

Reply:
This looks good - I think steps 1-3 are perfect to confirm this doesn't log out existing sessions if other users didn't opt in to this new functionality.

To make (5) seamless without a logout, the users need to get a new `groups` aware session cookie sent back somehow. You are right, a cookie expiry wouldn't work (since that results in a new login).  But I think a cookie-refresh should do it (that triggers the refresh token to get sent and generate a new access/ID token). I'll double check the code path to confirm this results in a whole new session serialization (which maybe get groups?)

If it isn't graceful, we can settle for a logout.  That is the secure route.

Reply:
Ok - glad we talked through this. There's a security issue we need to fix here: https://github.com/oauth2-proxy/oauth2-proxy/blob/d7abd56981c72ab71baf5429e5eefc7406c9677f/providers/oidc.go#L122

We need to make sure to update the `SessionState.Groups` field with any potential updates to the `groups` claim on the fresh IDToken after a refresh.

Otherwise you could strip a user's Group in the IdP assuming you'd be stripping access but the Groups would persist from the original login and not get updated.

Reply:
@NickMeves I beleive I am already doing this in this PR? line 127 in oidc.go.

```
if newSession.IDToken != "" {
    s.IDToken = newSession.IDToken
    s.Email = newSession.Email
    s.User = newSession.User
    s.Groups = newSession.Groups
    s.PreferredUsername = newSession.PreferredUsername
}
```

Reply:
Yep -- you are totally right. My bad!

So doing setting a low `refresh-cookie` timer to get a refresh didn't result in a new session that had groups that would pass your new group authZ check in step (5) above?

Reply:
So I tried that out, I set `--cookie-refresh=10s` and performed the following:

1. Deploy v6.0.0 and login getting a cookie
2. Deploy my branch refresh a number of times over the next minute
3. Deploy the groups restriction

After doing that I was logged out and had to go through the login flow again. I am using Dex with GitHub setup as my connector FWIW.

Reply:
Are you able to inspect your various Tokens & Headers at the upstream? IE hooking in `httpbin` (your own image or `httpbin.org/anything`) to confirm the IDToken claims you want are there and the `X-Forwarded-Groups` comes through appropriately in these stages?

I wouldn't think adding the groups restriction would force a re-login. I'll find time to peek through the logic again this afternoon.

Reply:
@NickMeves headers and everything looks good, I am not sure how adding the groups restriction won't cause it not to logout, as once it sees the user is not in the group it gets to this part https://github.com/stefansedich/oauth2-proxy/blob/e064f8ed4cd21bf8652782114cc4acf3749441b3/oauthproxy.go#L855 where it then shows the login page?

Reply:
My hope was if we enabled groups (and not restrict) in phase 1, upon refreshing my session might have `Groups: Foobar` or something.

Then upon updating the config to restrict by members of `Foobar` my session would pass this check because I have the group: https://github.com/stefansedich/oauth2-proxy/blob/e064f8ed4cd21bf8652782114cc4acf3749441b3/oauthproxy.go#L847

I guess this isn't even limited to 6.0.0 -> this Branch conversion.  I would hope that sessions that already have groups wouldn't be forced to reauthenticate if new group membership restrictions were added as a config option and my session met the restrictions.

Reply:
Does this reauth happen even during shifts? IE I'm a member of `Group1` & `Group2` and my session reflects that. And intial config has AuthZ restriction of `Group1`

If we switched the config to `Group2` instead of `Group1` -- we shouldn't be requiring a user who is a member of both to relogin 🤔

Reply:
@NickMeves so it handles changing the groups fine, I had it running restricted to `groupa` and then got a session, changed it to restrict to `groupb` which I am also a member of and everything kept working without logging me out.

What I did not see however was the session refresh under me suddenly showing groups, it is possible I did not test this properly so happy to be shown it working.

Reply:
I wonder if this is your problem with your OIDC provider: https://github.com/oauth2-proxy/oauth2-proxy/blob/52c7fe5b3ebf0e7da2bd264102a18ff31695218b/providers/oidc.go#L110

Perfectly feasible to get `nil, nil` back from that. Which would make only these 4 fields on the session change:
https://github.com/oauth2-proxy/oauth2-proxy/blob/52c7fe5b3ebf0e7da2bd264102a18ff31695218b/providers/oidc.go#L168

Hence no group update with the newly added groups support on the refresh redeem.

In which case - there can't be a graceful transition for those users in that scenario and we have this as good as it will get.